engineioxide/src/payload.rs

```diff
@@ -0,0 +1,156 @@
+use std::{io::BufRead, vec};
+
+use crate::{service::ProtocolVersion, errors::Error};
+
+pub const PACKET_SEPARATOR: u8 = b'\x1e';
+
+/// A payload is a series of encoded packets tied together.
+/// How packets are tied together depends on the protocol.
+pub struct Payload<R: BufRead> {
+    reader: R,
+    buffer: Vec<u8>,
+    #[allow(dead_code)]
+    protocol: ProtocolVersion,
+}
+
+type Item = Result<String, Error>;
+
+impl<R: BufRead> Payload<R> {
+    pub fn new(protocol: ProtocolVersion, data: R) -> Self {
+        Payload {
+            reader: data,
+            buffer: vec![],
+            protocol,
+        }
+    }
+
+    #[cfg(feature = "v3")]
+    fn next_v3(&mut self) -> Option<Item> {
+        match self.reader.read_until(b':', &mut self.buffer) {
+            Ok(bytes_read) => {
+                if bytes_read > 0 {
+                    // remove trailing separator
+                    if self.buffer.ends_with(&[b':']) {
+                        self.buffer.pop();
+                    }
+
+                    let buffer = std::mem::take(&mut self.buffer);
+                    let length = match String::from_utf8(buffer) {
+                        Ok(s) => {
+                            if let Ok(l) = s.parse::<usize>() {
+                                l
+                            } else {
+                                return Some(Err(Error::InvalidPacketLength));
+                            }
+                        },
+                        Err(_) => return Some(Err(Error::InvalidPacketLength)),
+                    };
+
+                    self.buffer.resize(length, 0);
+
+                    match self.reader.read_exact(&mut self.buffer) {
+                        Ok(_) => {
+                            let buffer = std::mem::take(&mut self.buffer);
+                            Some(String::from_utf8(buffer).map_err(Into::into))
+                        },
+                        Err(e) => Some(Err(Error::Io(e))),
+                    }
+                } else {
+                    None
+                }
+            }
+            Err(e) => Some(Err(Error::Io(e))),
+        }
+    }
```

Comment:
```suggestion
    fn next_v3(&mut self) -> Option<Item> {
        match self.reader.read_until(b':', &mut self.buffer) {
            Ok(bytes_read) => (bytes_read > 0).then(|| {
                if self.buffer.ends_with(&[b':']) {
                    self.buffer.pop();
                }

                let buffer = std::mem::take(&mut self.buffer);
                let length = String::from_utf8(buffer)
                    .map_err(|_| Error::InvalidPacketLength)
                    .and_then(|s| s.parse::<usize>().map_err(|_| Error::InvalidPacketLength))?;

                self.buffer.resize(length, 0);

                self.reader.read_exact(&mut self.buffer)?;
                let buffer = std::mem::take(&mut self.buffer);
                String::from_utf8(buffer).map_err(Into::into)
            }),
            Err(e) => Some(Err(Error::Io(e))),
        }
    }
```
using `then` here simplifies the flow avoiding Some/None conversion, what do you think about it?

Reply:
Wow much better.

Reply:
You can also do the same with `next_v4` function

Reply:
Agree (+ the str conversion for packet length rather than String)